File: solutions/JavaScript/top_k_frequent_max_heap.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
function topKFrequent(nums, k) {
    const freq = new Map();
    for (const num of nums) {
        freq.set(num, (freq.get(num) ?? 0) + 1);
    }
    // Sort entries by frequency descending and take first k elements
    return [...freq.entries()]
        .sort((a, b) => b[1] - a[1])
        .slice(0, k)
        .map(([num]) => num);
}

File: solutions/Rust/top_k_frequent_max_heap.rs
use std::collections::HashMap;

impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut freq: HashMap<i32, i32> = HashMap::new();
        for num in &nums {
            *freq.entry(*num).or_insert(0) += 1;
        }

        let mut pairs: Vec<(i32, i32)> = freq.into_iter().map(|(num, cnt)| (cnt, num)).collect();
        // Sort by frequency descending
        pairs.sort_unstable_by(|a, b| b.0.cmp(&a.0));

        pairs.into_iter().take(k as usize).map(|(_, num)| num).collect()
    }
}
